index.coffee

```diff
@@ -1,60 +1,161 @@
-through = require 'through'
+stream = require 'readable-stream'
+throttle = require 'lodash.throttle'
 
 module.exports = (requestModule, requestOpts, opts) ->
-	res = new Resumable(requestModule, requestOpts, opts)
-	return res.resume()
-
-class Resumable
-	constructor: (requestModule, requestOpts, opts) ->
-		@stream = through (data) =>
-			@bytesRead += data.length
-			@stream.emit('data', data)
+	return new ResumableRequest(requestModule, requestOpts, opts)
+
+module.exports.defaults = (defaults = {}) ->
+	(requestModule, requestOpts, opts) ->
+		opts = Object.assign {}, defaults, opts
+		return new ResumableRequest(requestModule, requestOpts, opts)
+
+cloneResponse = (response) ->
+	# creates an `http.IncomingMessage`-like stream, that can also be piped to.
+	s = new stream.PassThrough()
+	[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) ->
+		s[prop] = response[prop]
+	return s
+
+wrapLegacyStream = (s) ->
+	new stream.Readable().wrap(s)
+
+class ResumableRequest extends stream.Readable
+	constructor: (@requestModule, @requestOpts, opts = {}) ->
+		super()
+
+		@error = null
+		@request = null
+		@response = null
+
 		@bytesRead = 0
+		@bytesTotal = null
 		@retries = 0
-		@maxRetries = opts?.maxRetries ? 10
-		@contentLength = null
-		@requestOpts = requestOpts
-		@requestModule = requestModule
-		@error = null
+		@maxRetries = opts.maxRetries ? 10
+
+		@_retry = throttle(@_retry.bind(this), opts.retryInterval ? 1000)
+		@_reportProgress = throttle(@_reportProgress.bind(this), opts.progressInterval ? 1000)
+
+		@on 'pipe', =>
+			# request can be written to when uploading a resource to a URL.
+			# We do not support that -- resumable request can only be used
+			# to *download* a resource, i.e. only the response is resumable.
+			# piping will fail with a not-so-obvious error as soon as the stream
+			# gets going because we don't implement Writable, so it's detect this
+			# and emit a more appropriate error.
+			@error = new Error('ResumableRequest is not writable')
+			# have to abort asynchronously to allow client code to register 'error' handlers
+			process.nextTick =>
+				@_retry()
+
+		@_request()
+
+	abort: ->
+		@emit('abort')
+		@_end()
+
+	push: (data, encoding) ->
+		@retries = 0 # we got some data, reset number of retries
+		@bytesRead += data.length if data?
+		@_reportProgress()
+		return super(data, encoding)
+
+	_end: ->
+		return if @ended
+		@ended = true
+		doEnd = =>
+			@request.abort()
+			@_retry.cancel()
+			@_reportProgress()
+			@_reportProgress.flush()
+			@_reportProgress.cancel()
+			@emit('end')
+			if not @error?
+				@emit('complete')
+		if @response?
+			@response.end(doEnd)
+		else
+			doEnd()
+
+	_read: -> # noop -- we're manually pushing buffers into self
```

Comment:
@jhermsmeier it seems the stream never closes if I do that -- I did it similar to this: https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding. Thing is, request.js does not implement the stream API properly -- it's an "old style" stream. It has no `read` method and it doesn't send a `readable` event either. So I tried to support backpressure by pausing/resuming the response, but it causes request.js to error out with a timeout error. It's possible I'm not testing properly, but in any case, *I think* backpressure is handled automatically for us by the PassThrough stream we create out of the initial response [1], and the Readable stream we wrap the responses with [2]. Am I wrong thinking that?

[1] https://github.com/resin-io-modules/resumable-request/blob/cf9507670aac420ae10b2f86b78804513642286e/index.coffee#L135
[2] https://github.com/resin-io-modules/resumable-request/blob/cf9507670aac420ae10b2f86b78804513642286e/index.coffee#L140